src/rfc3161_client/verify.py

```diff
@@ -12,170 +10,328 @@
 
-    p7 = tsp_response.time_stamp_token()
-    try:
-        verify_signed_data(p7, verification_certificate)
-    except ValueError as e:
-        msg = f"Error while verifying certificates: {e}"
-        raise VerificationError(msg)
+        if self._roots:
+            verification_certificate.update(cert.public_bytes(Encoding.DER) for cert in self._roots)
 
-    return True
+        if self._intermediates:
+            verification_certificate.update(
+                cert.public_bytes(Encoding.DER) for cert in self._intermediates
+            )
 
+        p7 = tsp_response.time_stamp_token()
+        try:
+            self.verify_signed_data(p7, verification_certificate)
+        except ValueError as e:
+            msg = f"Error while verifying certificates: {e}"
+            raise VerificationError(msg)
 
-def verify_timestamp_response(
-    timestamp_response: TimeStampResponse, hashed_message: bytes, verify_opts: VerifyOpts
-) -> bool:
-    """Verify a Timestamp Response.
+        return True
 
-    Inspired by:
-        https://github.com/sigstore/timestamp-authority/blob/main/pkg/verification/verify.go#L209
+    def verify_signed_data(self, sig: bytes, certificates: set[bytes]) -> None:
+        """Verify signed data.
 
-    """
-    # Note: digitorus/timestamp does not validate if the result is GRANTED_WITH_MOD
-    # https://github.com/digitorus/timestamp/blob/master/timestamp.go#L268
-    if PKIStatus(timestamp_response.status) != PKIStatus.GRANTED:
-        msg = "PKIStatus is not GRANTED"
-        raise VerificationError(msg)
+        This function verifies that the bytes used in a signature are signed by a certificate
+        trusted in the `certificates` list.
+        The function does not return anything, but raises an exception if the verification fails.
 
-    _verify_tsr_with_chains(timestamp_response, verify_opts)
+        :param sig: Bytes of a PKCS7 object. This must be in DER format and will be unserialized.
+        :param certificates: A list of trusted certificates to verify the response against.
+        :raise: ValueError if the signature verification fails.
+        """
+        return _rust_verify.pkcs7_verify(sig, list(certificates))
 
-    # Verify Nonce
-    if verify_opts.nonce is not None and timestamp_response.tst_info.nonce != verify_opts.nonce:
-        msg = "Nonce mismatch"
-        raise VerificationError(msg)
 
-    # Verify Policy ID
-    if (
-        verify_opts.policy_id is not None
-        and timestamp_response.tst_info.policy != verify_opts.policy_id
-    ):
-        msg = "Policy ID mismatch"
-        raise VerificationError(msg)
+def create_verifier_from_request(
+    tsp_request: TimeStampRequest,
+    tsa_certificate: cryptography.x509.Certificate | None,
+    common_name: str | None,
+    root_certificates: list[cryptography.x509.Certificate] | None = None,
+    intermediates: list[cryptography.x509.Certificate] | None = None,
+) -> Verifier:
```

Comment:
Thinking out loud here, since the only fields we extract from the request are the `nonce` and the `policy`, this is mostly the same as `VerifyBuilder.__init__`, except we pass `tsp_request` instead of `(nonce, policy)`.

Maybe we could have a `VerifyBuilder.from_request(req)` instead?